nsv13/code/controllers/subsystem/starsystem.dm

```diff
@@ -86,17 +91,130 @@ Returns a faction datum by its name (case insensitive!)
 		if(F.id == id)
 			return F
 
-
 /datum/controller/subsystem/star_system/proc/add_blacklist(what)
 	enemy_blacklist += what
 	if(locate(what) in enemy_types)
 		enemy_types -= what
 
-/datum/controller/subsystem/star_system/proc/instantiate_systems()
+/datum/controller/subsystem/star_system/proc/instantiate_systems(_source_path = SSmapping.config.starmap_path)
+	message_admins("Loading starsystem from [_source_path]...")
+	var/list/_systems = list()
+	//Read the file in...
+	try
+	{
+		_systems = json_decode(rustg_file_read(file(_source_path)))
+	}
+
+	catch(var/exception/ex){
+		//Fallback: Load the hardcoded systems and report an error.
+		instantiate_systems_backup()
+		CRASH("FATAL: Unable to load starmap from: [_source_path]. (Defaulting...): [ex]")
+	}
+
+	for(var/i = 1; i <= _systems.len; i++)
+		//Try instancing this system from JSON, jump out if anything goes wrong.
+		var/list/sys_info = _systems[i]
+		try{
+			//Fields that must be present are accessed unsafely. The rest is fine if null.
+			var/datum/star_system/next = new /datum/star_system(
+				//Required props, you must fill these out, or it's an invalid map.
+				name = sys_info["name"],
+				desc = sys_info["desc"],
+				x = sys_info["x"],
+				y = sys_info["y"],
+				alignment = sys_info["alignment"],
+				hidden = sys_info["hidden"],
+				sector = sys_info["sector"],
+				adjacency_list = json_decode(sys_info["adjacency_list"]),
+				//Optional props. Recommended, but can be left blank.
+				threat_level = LAZYACCESS(sys_info, "threat_level") || THREAT_LEVEL_NONE,
+				is_capital = LAZYACCESS(sys_info, "is_capital") || FALSE,
+				fleet_type = LAZYACCESS(sys_info, "fleet_type") || null,
+				parallax_property = LAZYACCESS(sys_info, "parallax_property") || null,
+				visitable = LAZYACCESS(sys_info, "visitable") || TRUE,
+				is_hypergate = LAZYACCESS(sys_info,"is_hypergate") || FALSE,
+				preset_trader = (LAZYACCESS(sys_info,"preset_trader")) ? text2path(sys_info["preset_trader"]) : null,
+				system_traits = LAZYACCESS(sys_info,"system_traits") || list(),
+				system_type = (LAZYACCESS(sys_info,"system_type") && sys_info["system_type"] != "null" && sys_info["system_type"] != null) ? json_decode(sys_info["system_type"]) : list(),
+				audio_cues = (LAZYACCESS(sys_info,"audio_cues") && sys_info["audio_cues"] != "null" && sys_info["audio_cues"] != null) ? json_decode(sys_info["audio_cues"]) : list(),
+				wormhole_connections = (LAZYACCESS(sys_info,"wormhole_connections") && sys_info["wormhole_connections"] != "null" && sys_info["wormhole_connections"] != null) ? json_decode(sys_info["wormhole_connections"]) : list(),
+				startup_proc = LAZYACCESS(sys_info,"startup_proc") || null
```

Comment:
you only need lazy access if you're using a lazy list (a list where you're unsure if it's initialized or not), checking it's existence and length every access just adds unneeded overhead

Reply:
lazyaccess => l.hasKey(x) ? l[x] : null.
If the starmap designer omits keys, it's valid behaviour for these.

(Was my thinking)

Reply:
They're not going to runtime if your key search doesn't find anything/returns null. You'd use `LAZYACCESS` if you weren't sure if `sys_info` exists or not, it doesn't change any behavior relating to the key

Reply:
This is true for all lazy defines (that I know of), their main function is to dynamically initialize and check the existences of lists to save memory.